java/src/main/native/src/row_conversion.cu

```diff
@@ -447,32 +1073,558 @@ static inline int32_t compute_fixed_width_layout(std::vector<cudf::data_type> co
   }
 
   // Now we need to add in space for validity
-  // Eventually we can think about nullable vs not nullable, but for now we will just always add it
-  // in
-  int32_t validity_bytes_needed = (schema.size() + 7) / 8;
+  // Eventually we can think about nullable vs not nullable, but for now we will just always add
+  // it in
+  int32_t validity_bytes_needed =
+      (schema.size() + 7) / 8; // cudf::util::div_rounding_up_safe<int32_t>(schema.size(), 8);
   // validity comes at the end and is byte aligned so we can pack more in.
   at_offset += validity_bytes_needed;
   // Now we need to pad the end so all rows are 64 bit aligned
   return align_offset(at_offset, 8); // 8 bytes (64 bits)
 }
 
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
+
+/**
+ * @brief Compute information about a table such as bytes per row and offsets.
+ *
+ * @tparam iterator iterator of column schema data
+ * @param begin starting iterator of column schema
+ * @param end ending iterator of column schema
+ * @param column_starts column start offsets
+ * @param column_sizes size in bytes of each column
+ * @return size of the fixed_width data portion of a row.
+ */
+template <typename iterator>
+static size_type compute_column_information(iterator begin, iterator end,
+                                            std::vector<size_type> &column_starts,
+                                            std::vector<size_type> &column_sizes) {
+  size_type fixed_width_size_per_row = 0;
+  for (auto cv = begin; cv != end; ++cv) {
+    auto col_type = std::get<0>(*cv);
+    bool nested_type = col_type.id() == type_id::LIST || col_type.id() == type_id::STRING;
```

Comment:
Seems like you probably want `cudf::is_compound()` here, which covers strings, lists, structs and dictionaries.